### PR TITLE
Feat/verify signed data for issue endpoints

### PR DESCRIPTION
**Pull request by Judge40**
Use the Spring security filter chain to verify that requests to issue credentials (via `/api/issue/*` endpoints) are providing signed data. See commit 423d39522227dbde61e5ea8b8e311a36aaad48df for full details about the approach.

TIS21-4067

### CODE REVIEW

File: src/main/java/uk/nhs/hee/tis/trainee/credentials/api/IssueResource.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+public class IssueResource {
+
+  @PostMapping("/test")
+  ResponseEntity<String> issueTestCredential(@RequestBody TestCredential testCredential) {
+    return ResponseEntity.ok(
+        "Issuing Test Credential for %s %s".formatted(testCredential.givenName(),
+            testCredential.familyName));
+  }
+
+  private record TestCredential(String givenName, String familyName, LocalDate birthDate) {
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/hee/tis/trainee/credentials/api/IssueResource.java:45]: Nice, java 16 here we come :)

File: src/main/java/uk/nhs/hee/tis/trainee/credentials/filter/CachedBodyRequestWrapper.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    public boolean isReady() {
+      return true;
+    }
+
+    @Override
+    public void setReadListener(ReadListener listener) {
+      throw new UnsupportedOperationException();
+    }
+  }
+}
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/hee/tis/trainee/credentials/filter/CachedBodyRequestWrapper.java:103]: Ok, this looks like a fairly direct implementation of https://www.baeldung.com/spring-reading-httpservletrequest-multiple-times, which seems reasonable.

File: src/main/java/uk/nhs/hee/tis/trainee/credentials/filter/SignedDataFilter.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    this.mapper = mapper;
+    this.signatureSecretKey = signatureSecretKey;
+  }
+
+  @Override
+  protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response,
+      FilterChain filterChain) throws ServletException, IOException {
+    request = new CachedBodyRequestWrapper(request);
+
+    try {
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/hee/tis/trainee/credentials/filter/SignedDataFilter.java:70]: I wonder if a 403 is appropriate in all cases? I'm thinking a no-data or not-json request could possibly trigger a 400 instead. It probably doesn't matter much, but I thought worth flagging-up for consideration.

Comment (Judge40) [reply src/main/java/uk/nhs/hee/tis/trainee/credentials/filter/SignedDataFilter.java:70]: I want to look at returning a more reasonable error message, but I think it has to be implemented using interceptors so kept it simple to start with.

Comment (ReubenRobertsHEE) [reply src/main/java/uk/nhs/hee/tis/trainee/credentials/filter/SignedDataFilter.java:70]: Ok, fair enough 👍

**Review by ReubenRobertsHEE — APPROVED** [review]

**Pull request merged**